plancheck/expect_known_output_value.go

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package plancheck
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	tfjson "github.com/hashicorp/terraform-json"
+
+	"github.com/hashicorp/terraform-plugin-testing/knownvalue"
+	"github.com/hashicorp/terraform-plugin-testing/tfjsonpath"
+)
+
+// Resource Plan Check
+var _ PlanCheck = expectKnownOutputValue{}
+
+type expectKnownOutputValue struct {
+	outputAddress string
+	knownValue    knownvalue.KnownValue
+}
+
+// CheckPlan implements the plan check logic.
+func (e expectKnownOutputValue) CheckPlan(ctx context.Context, req CheckPlanRequest, resp *CheckPlanResponse) {
+	var change *tfjson.Change
+
+	for address, oc := range req.Plan.OutputChanges {
+		if e.outputAddress == address {
+			change = oc
+
+			break
+		}
+	}
+
+	if change == nil {
+		resp.Error = fmt.Errorf("%s - Output not found in plan OutputChanges", e.outputAddress)
+
+		return
+	}
+
+	result, err := tfjsonpath.Traverse(change.After, tfjsonpath.Path{})
+
+	if err != nil {
+		resp.Error = err
+
+		return
+	}
+
+	if result == nil {
+		resp.Error = fmt.Errorf("output value is null")
+
+		return
+	}
+
+	switch reflect.TypeOf(result).Kind() {
```

Comment:
We _might_ want to consider DRYing up the reflection logic as it's identical in `ExpectKnownValue`, `ExpectKnownOutputValue`, and `ExpectKnownOutputValueAtPath`.

Reply:
Reflection logic was substantially altered and reduced during the evolution of this PR.